dask/threaded.py

```diff
@@ -3,28 +3,53 @@
 
 See local.py
 """
-import atexit
 import multiprocessing.pool
 import sys
 import threading
-from collections import defaultdict
 from concurrent.futures import ThreadPoolExecutor
-from threading import Lock, current_thread
 
 from . import config
 from .local import MultiprocessingPoolExecutor, get_async
 from .system import CPU_COUNT
 from .utils_test import add, inc  # noqa: F401
 
+_EXECUTORS = threading.local()
 
-def _thread_get_id():
-    return current_thread().ident
+
+class _ExecutorPool:
+    """A thread-local pool of executors. When the thread closes, this pool will
+    be GC'd and all contained executors shutdown"""
+
+    def __init__(self):
+        self.executors = {}
+
+    def __del__(self):
+        for executor in self.executors.values():
+            executor.shutdown()
+
+    def get(self, num_workers):
+        if num_workers not in self.executors:
+            self.executors[num_workers] = ThreadPoolExecutor(num_workers)
+        return self.executors[num_workers]
+
+
+def get_executor(num_workers=None):
+    """Get a ThreadPoolExecutor given a specified number of workers.
+
+    Executors are local to their calling thread, and are cached between
+    subsequent calls. If the calling thread is closed, all executors will be
+    automatically cleaned up.
+    """
+    if not hasattr(_EXECUTORS, "pool"):
+        _EXECUTORS.pool = _ExecutorPool()
+    if not num_workers:  # treat both 0 and None the same
+        # TODO: if num_workers is 1 should we still use more threads?
```

Comment:
Right now we default to `num_workers == CPU_COUNT`. Do we want to set a minimum threshold here? This came up in a binder session where `CPU_COUNT` was `1`, so the threaded scheduler wasn't resulting in any parallelism on a simple `time.sleep` example.

`multiprocessing.pool.ThreadPool` defaults to  `os.cpu_count()`, but `concurrent.futures.ThreadPoolExecutor` does set a min default of `cpu_count + 4`.

Reply:
I was wondering since we are using the `concurrent.futures.ThreadPoolExecutor` shouldn't we try to be consistent with their defaults? 
Checking the docs when `max_workers=None` since version 3.8 it sets the default to `min(32, os.cpu_count() + 4)` and when `max_workers=0` it raises a value error, instead of treating it same as `None`. 

See: https://github.com/python/cpython/blob/9ccdc90488302b212bd3405d10dc5c22052e9b4c/Lib/concurrent/futures/thread.py#L128-L138

Reply:
Note we should still be using the custom [`dask.system.cpu_count`](https://github.com/dask/dask/blob/c5633c21e754df6b2c13195f1e2688e347c48963/dask/system.py#L13) function rather than `os.cpu_count` directly, as it handles things like cpu affinity and cgroups.

Reply:
IMHO using `CPU_COUNT` as a default still makes sense.

Likely the instances Binder uses lack multiple cores (and likely live in a multitenant deployment). If there is a better value to use there, it would make sense to document in notebooks or other docs so users know how to override it.

We could potentially check an environment variable and use that for `CPU_COUNT` (if Binder folks would like to override this generally). Have done similar things with conda-build and CI in conda-forge, which has worked well.

One other thought on the use of a single thread in the multithreaded case, we could issue a warning so the user is aware. This likely requires some thought to ensure the warning is not overly noise (or entirely missed).

Reply:
This is already settable with an environment variable (`DASK_NUM_WORKERS`), but not every user knows that.

I don't think we need to match the heuristics of other python threadpool-like-things, I was mainly wondering if having a threaded scheduler that may end up using only 1 thread by default on some systems would be confusing to new users. But since dask is mainly used for compute-heavy work (rather than IO heavy work), using more threads > cores outside of a demo context makes less sense. Fine to keep things as is.

Reply:
Ah right. Good point.

Yeah I think it is ok to leave as-is. We could maybe have a warning, but we could also handle that as a separate issue ( https://github.com/dask/dask/issues/8152 ).